### PR TITLE
ENH: Refactor of how spatial normalization targets and ``--output-spaces`` are maintained

### PR DESCRIPTION
**Pull request by oesteban**
Supersedes (and largely based on) #1915.

## Distinguish user-defined `--output-spaces` from internal intermediary steps
This PR ensures that spaces not requested (i.e. `fsaverage5` when running with freesurfer, `fsLR` surfaces when using `--cifti-output`, etc) are not included in the final output.

To do so, it utilizes the recently added infrastructure for `Space`s and `SpatialReferences`.

To avoid name collisions of the outputs, it uses a hack on our current DerivativesDataSinks. Those will need to be revised with the BIDS-Derivs spec because `cohort` and `res` appear at the end before the suffix (that's a side effect of the hack). But I guess this is better than not allowing it or rewriting silently.

### CODE REVIEW

File: .circleci/ds210_outputs.txt [hunk 1]
```diff
@@ -28,10 +28,5 @@ fmriprep/sub-02/anat/sub-02_space-MNI152NLin2009cAsym_label-WM_probseg.nii.gz
 fmriprep/sub-02/func
 fmriprep/sub-02/func/sub-02_task-cuedSGT_run-01_desc-confounds_regressors.json
 fmriprep/sub-02/func/sub-02_task-cuedSGT_run-01_desc-confounds_regressors.tsv
-fmriprep/sub-02/func/sub-02_task-cuedSGT_run-01_space-MNI152NLin2009cAsym_boldref.nii.gz
-fmriprep/sub-02/func/sub-02_task-cuedSGT_run-01_space-MNI152NLin2009cAsym_desc-brain_mask.json
-fmriprep/sub-02/func/sub-02_task-cuedSGT_run-01_space-MNI152NLin2009cAsym_desc-brain_mask.nii.gz
-fmriprep/sub-02/func/sub-02_task-cuedSGT_run-01_space-MNI152NLin2009cAsym_desc-preproc_bold.json
-fmriprep/sub-02/func/sub-02_task-cuedSGT_run-01_space-MNI152NLin2009cAsym_desc-preproc_bold.nii.gz
```
[the remark this answers is not in the document]
Comment (oesteban) [reply .circleci/ds210_outputs.txt]: Yup, this particular is WIP.

File: fmriprep/workflows/base.py [hunk 2]
```diff
@@ -88,6 +85,7 @@ def init_fmriprep_wf(
             from collections import namedtuple, OrderedDict
             BIDSLayout = namedtuple('BIDSLayout', ['root'])
             from fmriprep.workflows.base import init_fmriprep_wf
+            from niworkflows.utils.spaces import Space, SpatialReferences
```
Comment (effigies) [inline fmriprep/workflows/base.py]:
```suggestion
            from fmriprep.utils.spaces import Space, SpacesManager
```

File: fmriprep/workflows/base.py [hunk 3]
```diff
@@ -111,15 +109,16 @@ def init_fmriprep_wf(
...
-                    ('MNI152Lin', {}), ('fsaverage', {'density': '10k'}),
-                    ('T1w', {}), ('fsnative', {})]),
                 regressors_all_comps=False,
                 regressors_dvars_th=1.5,
                 regressors_fd_th=0.5,
                 run_uuid='X',
                 skull_strip_fixed_seed=False,
-                skull_strip_template=('OASIS30ANTs', {}),
+                skull_strip_template=Space.from_string('OASIS30ANTs')[0],
+                spaces=SpatialReferences([('MNI152Lin',),
```
Comment (effigies) [inline fmriprep/workflows/base.py]:
```suggestion
                spaces=SpacesManager([('MNI152Lin',),
```

File: fmriprep/workflows/base.py (hunk 2 above)
Comment (oesteban) [reply fmriprep/workflows/base.py]: Working on these RN - maybe best hold up until I push (within the next 20min)

Comment (effigies) [reply fmriprep/workflows/base.py]: Sounds good. Sorry. This was a lot to read through, so it's taken me like 2 hours of intermittent concentration.

I think I have a grasp on most of the moving parts, and the good news is I think things are overall getting simpler.

Comment (oesteban) [reply fmriprep/workflows/base.py]: > This was a lot to read through, so it's taken me like 2 hours of intermittent concentration.

Totally, I'll tell you how I call this refactor in private :p

> I think things are overall getting simpler.

This is important, good to know :)

File: .circleci/ds210_outputs.txt (hunk 1 above)
[the remark this answers is not in the document]
Comment (oesteban) [reply .circleci/ds210_outputs.txt]: We were actually not passing it to the full-fMRIPrep run. Should be fixed in next commit.

File: fmriprep/workflows/bold/resampling.py [hunk 4]
```diff
@@ -712,11 +640,175 @@ def init_bold_preproc_report_wf(mem_gb, reportlets_dir, name='bold_preproc_repor
...
+    from niworkflows.interfaces.cifti import GenerateCifti
+    workflow = Workflow(name=name)
+    workflow.__desc__ = """\
+*Grayordinates* files [@hcppipelines] containing {density} samples were also
+generated using the highest-resolution ``fsaverage`` as intermediate standardized
+surface space.
+""".format(density=grayord_density)
+
+    fslr_density = ('59k', '32k')[grayord_density == '91k']
+    mni_density = '32'[grayord_density == '91k']
```
Comment (mgxd) [inline fmriprep/workflows/bold/resampling.py]:
```suggestion
    mni_density = '12'[grayord_density == '91k']
```

File: fmriprep/workflows/bold/resampling.py [hunk 5]
```diff
@@ -712,11 +640,175 @@ def init_bold_preproc_report_wf(mem_gb, reportlets_dir, name='bold_preproc_repor
...
+    import templateflow.api as tf
+    from niworkflows.interfaces.cifti import GenerateCifti
+    workflow = Workflow(name=name)
+    workflow.__desc__ = """\
+*Grayordinates* files [@hcppipelines] containing {density} samples were also
+generated using the highest-resolution ``fsaverage`` as intermediate standardized
+surface space.
+""".format(density=grayord_density)
+
+    fslr_density = ('59k', '32k')[grayord_density == '91k']
```
Comment (oesteban) [inline fmriprep/workflows/bold/resampling.py]: for consistency with below:
```suggestion
    fslr_density, mni_density = ('32k', '2') if grayord_density == '91k' else ('59k', '1')
```

File: fmriprep/workflows/bold/resampling.py (hunk 4 above)
Comment (oesteban) [inline fmriprep/workflows/bold/resampling.py]: if done above,
```suggestion
```

File: fmriprep/workflows/bold/resampling.py (hunk 5 above)
Comment (mgxd) [reply fmriprep/workflows/bold/resampling.py]: this looks much cleaner

File: fmriprep/workflows/bold/resampling.py [hunk 6]
```diff
@@ -397,21 +337,7 @@ def init_bold_std_trans_wf(
...
-        (select_std, select_tpl, [('key', 'template')]),
-        (select_std, mask_merge_tfms, [('anat2std_xfm', 'in1')]),
-        (select_std, gen_ref, [(('resolution', _is_native), 'keep_native')]),
-        (select_tpl, gen_ref, [('out', 'fixed_image')]),
-        (mask_merge_tfms, mask_std_tfm, [('out', 'transforms')]),
-        (gen_ref, mask_std_tfm, [('out_file', 'reference_image')]),
-    ])
-
-    nxforms = 4 if use_fieldwarp else 3
+    nxforms = [3, 4][use_fieldwarp]
```
[the remark this answers is not in the document]
Comment (mgxd) [reply fmriprep/workflows/bold/resampling.py]: i'll leave it up to you

**Comment by oesteban** [comment]
This PR interacts with several issues:
  - Fixes #1739 
  - Fixes #1810 
  - Implements and closes #1817

File: setup.cfg [hunk 7]
```diff
@@ -24,14 +24,14 @@ install_requires =
...
-    niworkflows ~= 1.1.3
+    niworkflows ~= 1.1.5
     numpy
     pandas
     psutil >=5.4
     pybids ~= 0.9.4
     pyyaml
     sdcflows ~=1.1.0
-    smriprep ~= 0.4.2
+    smriprep @ git+https://github.com/poldracklab/smriprep.git@46270e8847f36a2dd677e810162358eb69acb211
```
Comment (mgxd) [inline setup.cfg]:
```suggestion
    smriprep ~= 0.5
```

Comment (effigies) [reply setup.cfg]: 0.5.0. the distinction matters with ~=

Comment (effigies) [reply setup.cfg]: Oh, looks like Oscar already did this. (On my phone...)

**Pull request merged**